engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala

```diff
@@ -7206,6 +7206,272 @@ class RuntimeServerTest
+    context.receive shouldEqual Some(
+      Api.Response(requestId, Api.CreateContextResponse(contextId))
+    )
+
+    // open file
+    context.send(
+      Api.Request(requestId, Api.OpenFileRequest(mainFile, contents))
+    )
+    context.receive shouldEqual Some(
+      Api.Response(Some(requestId), Api.OpenFileResponse)
+    )
+
+    // push main
+    context.send(
+      Api.Request(
+        requestId,
+        Api.PushContextRequest(
+          contextId,
+          Api.StackItem.ExplicitCall(
+            Api.MethodPointer(moduleName, moduleName, "main"),
+            None,
+            Vector()
+          )
+        )
+      )
+    )
+    context.receiveNIgnoreStdLib(4) should contain theSameElementsAs Seq(
+      Api.Response(requestId, Api.PushContextResponse(contextId)),
+      TestMessages.update(
+        contextId,
+        idX,
+        ConstantsGen.INTEGER,
+        methodCall =
+          Some(Api.MethodCall(Api.MethodPointer(moduleName, moduleName, "inc")))
+      ),
+      TestMessages.update(contextId, idSelfMain, moduleName),
+      context.executionComplete(contextId)
+    )
+    context.consumeOut shouldEqual List("42")
+
+    // push inc call
+    context.send(
+      Api.Request(
+        requestId,
+        Api.PushContextRequest(contextId, Api.StackItem.LocalCall(idX))
+      )
+    )
+    context.receiveNIgnoreStdLib(2) should contain theSameElementsAs Seq(
+      Api.Response(requestId, Api.PushContextResponse(contextId)),
+      context.executionComplete(contextId)
+    )
+    context.consumeOut shouldEqual List("42")
+
+    // Modify the file
+    context.send(
+      Api.Request(
+        Api.EditFileNotification(
+          mainFile,
+          Seq(
+            model.TextEdit(
+              model.Range(model.Position(8, 4), model.Position(8, 9)),
+              "z = 2\n    r = a + z"
+            )
+          ),
+          execute = true,
+          idMap   = Some(model.IdMap(Vector(model.Span(102, 103) -> idIncZ)))
+        )
+      )
+    )
+    context.receiveNIgnorePendingExpressionUpdates(2, 10) shouldEqual Seq(
+      TestMessages.update(
+        contextId,
+        idIncZ,
+        ConstantsGen.INTEGER
+      ),
+      context.executionComplete(contextId)
+    )
+    context.consumeOut shouldEqual List("44")
+  }
+
+  it should "edit two local calls with cached self" in {
+    val contextId  = UUID.randomUUID()
+    val requestId  = UUID.randomUUID()
+    val moduleName = "Enso_Test.Test.Main"
+
+    val metadata    = new Metadata
+    val idX         = metadata.addItem(46, 10, "aa")
+    val idY         = metadata.addItem(65, 10, "ab")
+    val idXSelfMain = metadata.addItem(46, 4, "ac")
+    val idYSelfMain = metadata.addItem(65, 4, "ad")
+    val idIncZ      = new UUID(0, 1)
+
+    val code =
+      """from Standard.Base import all
+        |
+        |main =
+        |    x = Main.inc 3
+        |    y = Main.inc 7
+        |    IO.println x+y
```

Comment:
It was not working because the `self` argument was not invalidated after the method was edited. With the cached `self` we observed some stale nodes during the runtime, and the function call may return incorrect results (the result before the edit was done). Why it was the behavior is another question and I was not able to find the answer.